segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/SegmentAttributeIndex.java

```diff
@@ -0,0 +1,663 @@
+                    }
+                }, this.executor);
+
+        if (!mustComplete) {
+            result = result.exceptionally(ex -> {
+                // Failure to update snapshot location in the main segment's metadata or to truncate the Attribute Segment
+                // is not a critical failure. The snapshot has been successfully written, and its data can be recovered.
+                // If we bubble up this exception, the caller may think the data have not been written and retry, hence
+                // unnecessarily write the data multiple times.
+                // If we do get here, we'll simply have to process more data when we read and potentially have more
+                // data than needed in the attribute segment, both of which will be fixed with the next snapshot attempt.
+                ex = Exceptions.unwrap(ex);
+                log.warn("{}: Snapshot serialized to attribute segment, but failed to update snapshot location or truncate Attribute Segment.", this.traceObjectId, ex);
+                return null;
+            });
+        }
+        return result;
+    }
+
+    /**
+     * Appends the result of the given serialization function conditionally based on the current length of the Attribute Segment.
+     * This method will retry the write subject to the APPEND_RETRY policy.
+     *
+     * @param getSerialization A Supplier that, when invoked, returns a CompletableFuture whose result will be the requested
+     *                         serialization.
+     * @param timer            Timer for the operation. Used for timeouts.
+     * @return A CompletableFuture that, when completed, will contain information about the data that was just written.
+     */
+    private CompletableFuture<WriteInfo> appendConditionally(Supplier<CompletableFuture<ArrayView>> getSerialization, TimeoutTimer timer) {
+        return APPEND_RETRY.runAsync(() -> appendConditionallyOnce(getSerialization, timer), this.executor);
+    }
+
+    /**
+     * Appends the result of the given serialization function conditionally based on the current length of the Attribute Segment.
+     * This method does not perform any retries.
+     *
+     * @param getSerialization A Supplier that, when invoked, returns a CompletableFuture whose result will be the requested
+     *                         serialization.
+     * @param timer            Timer for the operation. Used for timeouts.
+     * @return A CompletableFuture that, when completed, will contain information about the data that was just written.
+     */
+    private CompletableFuture<WriteInfo> appendConditionallyOnce(Supplier<CompletableFuture<ArrayView>> getSerialization, TimeoutTimer timer) {
+        // We want to make sure that the serialization we generate is accurate based on the state of the Attribute Segment.
+        // This is to protect against potential corruptions due to concurrency: for example we picked data for a Snapshot,
```

Comment:
Do we have test cases to validate that corruption isn't possible due to concurrent access?

Reply:
Yes: `testConditionalUpdates`. That is the most we can test; we rely on writes to Tier2 to be atomic (i.e., if we write a sequence of N bytes, it will go in at once, and not be broken up).